scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,118 @@
+
+This approach needs to be revised for the following reasons:
+
+1.  The vulnerability schema rarely changes, but maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
+2.  It can be challenging to backport fixes and updates to previous releases.  This requires patch releases, which trigger the release workflow.  Improving the release process can be complex, as it may require patch release cuts, changes to different branches, and integration between logic that is managed in the master branch, executing other releases binaries.
+3.  The latest offline bundles does not support post `4.5` releases.  Bundle embedding for each release increases the cost to store and consume offline bundles.  The `roxctl` interface and documentation today needs to be clearer on these limitations.
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current StackRox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at the current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
+
+### Building versioned bundles
+
+The GitHub workflow will have a map with "bundle version" as keys and "git reference" as values, similar to Table 1 above.  This can be stored statically in the matrix definition.
+
+For each version in the mapping, a bundle will be built based on the specified git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.
```

Comment:
> It sounds like this is only updated upon release, but I think we'd want to update this as soon as we are aware of a schema change (sometime during the dev cycle). Is this allowed? If so, can we make it clear?

I thought it was clear. I will explain here and you let me know.

> When a schema bump is detected (e.g., a ClairCore bump that brings in incompatible schema changes to the vulnerability structs), `scanner/VULNERABILITY_VERSION` should be updated to the new schema version.

It means as soon as we are aware of a schema change (during the dev cycle) we bump the schema in `scanner/VULNERABILITY_VERSION`. It means the "dev" version bundle stream will be on the new schema, all non-release builds will break until they upgrade, but that's OK. Release and nightlies will continue to use "v1" until they upgrade.

> In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.

This is the mapping. It says "for the current VULNERABILITY_VERSION, what is the git ref we should use to build it?" -- we need to either bump it to the latest release, or create a new entry in the map to leave the last version in its current release.

I hope that clarifies.

Reply:
All good? Still have concerns?